### PR TITLE
major refactor (includes tons of documentation)

### PR DESCRIPTION
**Pull request by ExpandingMan**
I've been working on this for a little while now.  I've made quite a few changes, so I wanted to get this in front of people sooner rather than later as I've made a number of breaking changes.

**Please take a look at `docs/src/faq.md` for my justifications for changes!**

I expect that on first glance it will look like I changed a lot more here than I actually did.  I took the liberty of re-organizing the package as I felt it was desperately in need of cleaning up, but although I made tons of changes to organization and function text, the main functionality is much the same as it was after Keno's cursors PR was merged.  The cursors themselves, while reorganized (I consider them to now be much less confusing) don't work any differently.  Most of the changes to iteration are cosmetic, there are no breaking changes in the iterators.

I am fully aware that a huge number of packages depend on this.  This gave me a lot of motivation to clean this package up as it was feeling like quite a shaky foundation to me.  I realize that tagging 0.4 is going to involve a lot of work in the ecosystem.  I've already done a brief review of some of the dependents, and I do not expect most packages to need much in the way of changes if this is merged, and I fully intend to go on a PR spree bumping versions and making simple changes if we tag a 0.4.

Here is a list of tasks I still need to complete before it would make sense to merge this:
- [x] Finish deciding on exactly how `treemap` should work, checking that all previous use cases can be covered in a relatively straightforward way (see my FAQ).
- [x] Decide what, if anything, is needed to substitute the functionality of `ShadowTree`.  I'm not sure I fully understood the intended use cases.  (Doesn't look to me like anything should be done here, please let me know if you disagree.)
- [ ] Check the ecosystem for more complicated implementations to make sure they are still well-supported (my hope is that these will be much better off after the changes but they may require non-trivial updates).
- [ ] Finish documentation (it's mostly done, think it would be nice to have more complete examples).
- [ ] Make sure all existing unit tests have equivalents.
- [ ] New unit tests for new functionality.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name = "AbstractTrees"
 uuid = "1520ce14-60c1-5f80-bbc7-55ef81b5835c"
-version = "0.3.4"
+version = "0.3.5"
```
Comment (oscardssmith) [inline Project.toml:3]: 0.4.0?

Comment (ExpandingMan) [reply Project.toml:3]: Yeah, I just didn't do it yet because I didn't want Julia to complain about breaking dependencies to me lol... I'll change it soon

File: src/base.jl [hunk 2]
```diff
@@ -1,59 +1,149 @@
...
+By default, every object is a parent node of an empty set of children.  This is to make it simpler to define
+trees with nodes of different types, for example arrays are trees regardless of their `eltype`.
+
+**REQUIRED**: This is required for all tree nodes with non-empty sets of children.  If it is not possible to infer
+the children from the node alone, this should be defined for a wrapper object which does.
+"""
+children(node) = ()
+
+"""
+    siblings(node)
```
Comment (oscardssmith) [inline src/base.jl:48]: this should document that `x in siblings(x)`

File: src/base.jl [hunk 3]
```diff
@@ -1,59 +1,149 @@
...
+    nextsibling(node)
 
+Get the next sibling (child of the same parent) of the tree node `node`.  The returned node should be the same as
+the node that would be returned after `node` when iterating over `(children ∘ parent)(node)`.
 
+**OPTIONAL**: This function is required for nodes with the [`StoredSiblings`](@ref) trait.  There is no default
+definition.
 """
-    ischild(node1, node2)
+function nextsibling end
```
Comment (oscardssmith) [inline src/base.jl:69]: should this have a default definition for indexed trees?

File: src/base.jl [hunk 4]
```diff
@@ -1,59 +1,149 @@
...
+    ischild(node1, node2; equiv=(≡))
 
 Check if `node1` is a child of `node2`.
 
-By default this iterates through ``children(node2)``, so performance may be improved by adding a
+By default this iterates through `children(node2)`, so performance may be improved by adding a
 specialized method for given node type.
+
+Equivalence is established with the `equiv` function.  New methods of this function should include this
+argument or else it will fall back to `≡`.
```
Comment (oscardssmith) [inline src/base.jl:91]: Is `≡` the same as `===`? If so, I think you should document it as `===`.

File: src/base.jl [hunk 5]
```diff
@@ -1,59 +1,149 @@
...
-ischild(node1, node2) = any(node -> node === node1, children(node2))
+    parent(node)
+
+Get the immediate parent of a node `node`.
+
+By default all objects are considered nodes of a trivial tree with no children and no parents.  That is,
+the default method is simply `parent(node) = nothing`.
 
+**OPTIONAL**: The 1-argument version of this function must be implemented for nodes with the [`StoredParents`](@ref)
+trait.  The 2-argument version is always optional.
```
Comment (oscardssmith) [inline src/base.jl:104]: what 2 argument version?

File: src/cursors.jl [hunk 6]
```diff
@@ -1,270 +1,268 @@
...
-end
 
+"""
+    TreeCursor{P,N}
+
+Abstract type for tree cursors which when constructed from a node can be used to
+navigate the entire tree descended from that node.
+
+Tree cursors satisfy the abstract tree interface with a few additional guarantees:
+- In addition to [`children`](@ref), [`parent`](@ref), [`nextsibling`](@ref) and
```
Comment (oscardssmith) [inline src/cursors.jl:19]: Tree cursors must define [`children`](@ref), [`parent`](@ref), and [`nextsibling`](@ref). [`prevsibling`](@ref) may optionally be defined.

File: src/cursors.jl [hunk 7]
```diff
@@ -1,270 +1,268 @@
...
+    optionally [`prevsibling`](@ref) must be defined.
+- The above functions returning tree nodes are guaranteed to also return tree cursors.
```
Comment (oscardssmith) [inline src/cursors.jl:21]: what does this mean?

File: src/cursors.jl [hunk 8]
```diff
@@ -1,270 +1,268 @@
...
+
+Tree nodes which define `children` and have the traits [`StoredParents`](@ref) and
+[`StoredSiblings`](@ref) satisfy the `TreeCursor` interface and can be used as such.
```
Comment (oscardssmith) [inline src/cursors.jl:24]: does this mean that these tree nodes need to be `<:TreeCursor`?

File: src/cursors.jl [hunk 9]
```diff
@@ -1,270 +1,268 @@
...
-struct UnIndex{T} <: TreeCursor
-    tc::T
-end
-function nextsibling(ui::UnIndex)
-    ns = nextsibling(ui.tc)
-    ns === nothing && return nothing
-    UnIndex(ns)
+Get the type of the wrapped node.  This should match the return type of [`nodevalue`](@ref).
+"""
+nodevaluetype(::Type{<:TreeCursor{P,N}}) where {P,N} = N
```
Comment (oscardssmith) [inline src/cursors.jl:40]: why do we need the method taking the type? It seems to me that it would be simpler if we just defined `nodevaluetype(csr::TreeCursor{P,N}) where {P,N} = N`

File: src/cursors.jl [hunk 10]
```diff
@@ -1,270 +1,268 @@
...
+
+
+"""
+    TrivialCursor{P,N} <: TreeCursor{P,N}
+
+A [`TreeCursor`](@ref) which matches the functionality of the underlying node.  Tree nodes wrapped by this
+cursor themselves have most of the functionality required of a `TreeCursor`, this type exists entirely
+for the sake of maintaining a fully consistent interface with other `TreeCursor` objects.
+"""
+struct TrivialCursor{P,N} <: TreeCursor{P,N}
```
Comment (oscardssmith) [inline src/cursors.jl:80]: Is this basically the replacement of `UnIndex`?

File: src/iteration.jl [hunk 11]
```diff
@@ -1,46 +1,127 @@
...
+abstract type IteratorState{T<:TreeCursor} end
+
+# we define this explicitly to avoid run-time dispatch
+"""
+    instance(::Type{<:IteratorState}, node; kw...)
+
+Create an instance of the given [`IteratorState`](@ref) around node `node`.  This is mostly just a constructor
+for `IteratorState` except that if `node` is `nothing` it will return `nothing`.
+"""
+function instance(::Type{S}, node; kw...) where {S<:IteratorState}
```
Comment (oscardssmith) [inline src/iteration.jl:20]: this function feels weird to me.

File: src/iteration.jl [hunk 12]
```diff
@@ -1,46 +1,127 @@
...
+            csr = parent(csr)
+        else
+            return PreOrderState(n)
+        end
+    end
+    nothing
 end
-PostOrderDFS(tree::Tree) = PostOrderDFS(tree.x)
-IteratorSize(::Type{PostOrderDFS{T}}) where T = SizeUnknown()
+next(s::PreOrderState) = next(x -> true, s)
```
Comment (oscardssmith) [inline src/iteration.jl:119]: `next(_ -> true, s)`

File: src/iteration.jl [hunk 13]
```diff
@@ -1,46 +1,127 @@
...
+    cursor::T
+
+    PreOrderState(csr::TreeCursor) = new{typeof(csr)}(csr)
+end
+
+PreOrderState(node) = (PreOrderState ∘ TreeCursor)(node)
+
+initial(::Type{PreOrderState}, node) = PreOrderState(node)
+
+function next(f, s::PreOrderState)
```
Comment (oscardssmith) [inline src/iteration.jl:102]: This is really cool functionality.

File: src/iteration.jl [hunk 14]
```diff
@@ -1,46 +1,127 @@
...
+        isempty(ch) || return instance(PreOrderState, first(ch))
+    end
+
+    csr = s.cursor
+    while !isroot(csr)
+        n = nextsibling(csr)
+        if isnothing(n)
+            csr = parent(csr)
+        else
+            return PreOrderState(n)
```
Comment (oscardssmith) [inline src/iteration.jl:114]: doesn't this need a call to `f`?

File: src/iteration.jl [hunk 15]
```diff
@@ -59,93 +139,194 @@ Any[Any[1,2],Any[3,4]]
...
-function descend_left(cursor)
-    ccs = children(cursor)
-    isempty(ccs) && return cursor
-    return descend_left(first(ccs))
+
+PostOrderState(node) = (PostOrderState ∘ TreeCursor)(node)
+
+initial(::Type{PostOrderState}, node) = (PostOrderState ∘ descendleft ∘ TreeCursor)(node)
+
+function next(s::PostOrderState)
```
Comment (oscardssmith) [inline src/iteration.jl:177]: why no version with `f` here?

File: src/iteration.jl [hunk 16]
```diff
@@ -199,124 +360,123 @@ we will get `[[1, [2,3]], 1, [2, 3], 2, 3]`.
...
+
+Use [`AbstractTrees.nodevalue`](@ref) or `mapnode.value` to obtain the wrapped value.
+"""
+struct MapNode{T,C}
+    value::T
+    children::C
+
+    #TODO: need a way to modify children but hard to think of a good general way of doing it...
+    function MapNode(𝒻, node)
+        v = 𝒻(node)
```
Comment (oscardssmith) [inline src/iteration.jl:433]: `𝒻` renders a lot like `/` in some fonts.

File: src/printing.jl [hunk 17]
```diff
@@ -21,7 +21,7 @@ Print a text representation of `tree` to the given `io` object.
 
 # Examples
 
-```jldoctest; setup = :(using AbstractTrees)
+```julia
```
Comment (oscardssmith) [inline src/printing.jl:24]: can these printing changes go in a separate PR?

**Comment by oscardssmith** [comment]
I am currently in the process of trying to implement this for a fairly complicated (proprietary) usage. I'll let you know how it looks after.

File: src/base.jl (hunk 3 above)
Comment (ExpandingMan) [reply src/base.jl:69]: It does, via `IndexNode` (requires `nextsiblingindex`).

File: src/base.jl (hunk 2 above)
Comment (ExpandingMan) [reply src/base.jl:48]: I decided this version of `siblings` doesn't make a lot of sense in the first place, you can call `children(parent(node))`, so I removed it entirely.

File: src/base.jl (hunk 4 above)
Comment (ExpandingMan) [reply src/base.jl:91]: Yes, it is the same.  I have changed the text.

File: src/base.jl (hunk 5 above)
Comment (ExpandingMan) [reply src/base.jl:104]: No longer exists, text removed.

File: src/cursors.jl (hunk 9 above)
Comment (ExpandingMan) [reply src/cursors.jl:40]: Because the type is known at compile time whereas the value can only be known at run-time.

File: src/cursors.jl (hunk 10 above)
Comment (ExpandingMan) [reply src/cursors.jl:80]: I *think* so... don't think they are exactly the same, but basically.

File: src/cursors.jl [hunk 18]
```diff
@@ -1,270 +1,268 @@
...
+nodevaluetype(::Type{<:TreeCursor{P,N}}) where {P,N} = N
+nodevaluetype(csr::TreeCursor) = nodevaluetype(typeof(csr))
+
+"""
+    parenttype(csr::TreeCursor)
+
+The return type of `parent(csr)`.  For properly constructed `TreeCursor`s this is guaranteed to be another
+`TreeCursor`.
+"""
+parenttype(::Type{<:TreeCursor{P,N}}) where {P,N} = P
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/cursors.jl:49]: Again, type can be known at compile time.

File: src/cursors.jl [hunk 19]
```diff
@@ -1,270 +1,268 @@
...
+This should be thought of as a "worst case scenario" tree cursor.  In particular, `ImplicitCursor`s store the
+child iteration state of type `S` and for any of `ImplicitCursor`s method to be type-stable it must be possible
+to infer the child iteration state type, see [`childstatetype`](@ref).
+"""
+struct ImplicitCursor{P,N,S} <: TreeCursor{P,N}
+    parent::P
+    node::N
+    sibling_state::S
+
+    ImplicitCursor(p::Union{Nothing,ImplicitCursor}, n, s=InitialState()) = new{typeof(p),typeof(n),typeof(s)}(p, n, s)
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/cursors.jl:122]: `new` takes explicit type parameters... I *think* always.

File: src/cursors.jl [hunk 20]
```diff
@@ -1,270 +1,268 @@
+
 """
-    NodeCompletion
+    InitialState
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/cursors.jl:3]: It is needed because without it there is no way of specifying that the underlying iterator is yet to be called.  I suppose I could have used `Nothing` but I found that too weird because of the fact that `nothing` indicates the end of iteration.

File: src/iteration.jl (hunk 11 above)
Comment (ExpandingMan) [reply src/iteration.jl:20]: I can't use a constructor here because it doesn't necessarily return an object of type `S`.  This is not mandated by the compiler but I think it's a good convention.

File: src/iteration.jl (hunk 14 above)
Comment (ExpandingMan) [reply src/iteration.jl:114]: I don't see how... `f` is passed to `next` and called every time `next` is called.

File: src/iteration.jl (hunk 15 above)
Comment (ExpandingMan) [reply src/iteration.jl:177]: Because of the topology of the tree... skipping a parent implies you want to skip the children, but skipping a child does not necessarily imply you want to skip the parent (i.e. parents can have multiple children but children can only have one parent).

It was like this prior to my overhaul.

Something could be added here, but it's not clear to me what would be a good alternative, so if you think it should be added please let me know how you imagine this working.

File: src/iteration.jl [hunk 21]
```diff
@@ -155,32 +336,12 @@ function descend(select, tree)
...
+descend(select, node) = descend(select, ChildIndexing(node), node)
 
-"""
-    Base.searchsortedlast(leaves::Leaves, x; by=getnode)
 
-Return the last leaf of `leaves` less than or equal to `x`. Assumes that leaves
-are in sorted order and that non-leaf nodes sort equal to the smallest of their
-children.
 """
-function Base.searchsortedlast(leaves::Leaves, v; by=getnode)
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/iteration.jl:166]: I'll take a look at it again.

File: src/printing.jl (hunk 17 above)
Comment (ExpandingMan) [reply src/printing.jl:24]: Nope.  I think printing is ok, but adding the tests back in is something I still have to do.

Printing tests are extremely finicky so I intend on putting them in their on testset.

File: src/cursors.jl (hunk 8 above)
Comment (ExpandingMan) [reply src/cursors.jl]: Sorry, this was confusing because what I was describing was from a much earlier commit, I have changed the text.

File: src/iteration.jl (hunk 13 above)
Comment (ExpandingMan) [reply src/iteration.jl:102]: I'm glad it seems useful.  I was a bit nervous about how well-received `next` would be since it seems a little dubious to be nesting two different iteration protocols.  It's not a simplification to remove it so I thought I was better off doing it this way.

File: src/printing.jl (hunk 17 above)
Comment (oscardssmith) [reply src/printing.jl:24]: makes sense

**Comment by oscardssmith** [comment]
My initial impression is that this is pretty great, but will be a bunch of work to get the ecosystem to support this. (the tests I wanted to do have already required updating MTK and LeftChildRightSiblingTrees). The good news is so far the changes I've had to make are totally sensible.

**Comment by ExpandingMan** [comment]
Sounds good.  Like I said, I did a bit of quick checking to see how this might affect other packages, and thus far I am hopeful that it can go about as smoothly as updating the major version of such a prominent package can go due to the fact that most dependencies seem to mostly use `children` together with the iterators and the interface for this is unchanged in my update.  (The fact that so much of the original functionality was undocumented helps me here, since people weren't likely to use it.)

I plan on going on a PR spree to update dependencies, but my ability to do this is premised on the assumption that most dependencies use AbstractTrees in a relatively straightforward way.  If anyone knows of a particularly subtle or complicated example I'd like to see it so that I can test it.

I am going to check off the boxes in the opening comment of the PR as I progress, so check back there fore progress.

File: docs/src/faq.md [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+Iterators types do *not* have breaking changes.
+
+There are quite a few breaking changes for features of AbstractTrees.jl which were not documented or
+poorly documented and were therefore unlikely to be used.
+
+The most significant changes are for indexed trees which now rely on the [`IndexNode`](@ref) object
+and a dedicated set of methods.  Authors of packages using indexed trees should review [The Indexed
+Tree Interface](@ref).  Roughly speaking, the changes are
+- `children(tree, node)` ``\rightarrow`` `childindices(tree, node_index)`
+- `Iterator(tree)` ``\rightarrow`` `Iterator(IndexNode(tree))`
```
Comment (oscardssmith) [inline docs/src/faq.md:19]: should also mention `parentlinks`, `siblinglinks`, and the `eltype` changes.

File: src/cursors.jl [hunk 23]
```diff
@@ -1,270 +1,269 @@
...
+    idx > length(p) && return nothing
+    p[idx]
 end
 
-function Base.getindex(ltc::LinkedTreeCursor, idx)
-    cs = children(ltc.node)
-    LinkedTreeCursor(ltc, cs[idx], SiblingState(cs, idx))
+function prevsibling(csr::IndexedCursor)
+    idx = csr.index - 1
+    idx < 1 && return nothing
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/cursors.jl:211]: I think the name `IndexedCursor` is misleading here.  This has nothing to do with "indexed trees" which use `IndexNode`s, it merely assumes the trait `IndexedChildren`.

For a while I struggled with naming conventions as I wanted to distinguish `IndexNode` trees from `IndexedChildren`, for a while I was using the name `ReferenceNode` or `RefNode` and referred to it as a "reference tree", but I ultimately rejected this first because what's really going on with these trees has a lot more to do with Julia's indexing than references, and also because I was afraid of making it seem like there was too big a departure from v0.3 of this package.

I'm certainly open to changing some of the names here, but if you want to do this please make some suggestions, as I had a difficult time coming up with better alternatives.

Another issue is that I'm currently assuming `ChildIndexing` is 1-based.  This was pretty much just laziness, but considering the number of changes already in this PR, I would ask that we address this in a later PR (I don't think I'm breaking any functionality that previously existed for non-1-based indices).

Note that (as far as I know) the indexed tree interface does *NOT* assume 1-based indexing and should allow for *any* type of indices (not necessarily integers).

**Comment by ExpandingMan** [comment]
I have written a version of `treemap` that I am reasonably happy with.  I think the internals of it are correct, but I worry a little bit that the interface is confusing.

There are two options:
- `treemap(f, node)` takes a function `f` that returns `(newvalue, newchildren)`
- `treemap` takes 3 arguments, one being a function that takes `node` and returns a value, the other being a function that takes `node` and returns the children.

Ultimately I decided that the former is less confusing, but I'm open to changing it if you disagree (though perhaps try it out first).

There may be a slight foot-gun here in that you can e.g. cause a `StackOverflow` by always returning non-empty children, but this seems unavoidable regardless of the interface.  I tried to warn about this as best I could in the docs.

File: src/iteration.jl (hunk 21 above)
[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/iteration.jl:166]: I'll make a followup PR for this.

[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/iteration.jl:166]: The purpose of this function is that it is what you want to find an element in a sorted tree (eg b-tree) which comes up a decent amount.

**Comment by Keno** [comment]
I unfortunately don't have time for an in-depth review, but I skimmed it and I didn't see any major red flags. That said, we do have extensive use of AbstractTrees in private packages that I want to make sure keeps working (can be changed to new APIs of course, but the feature set should remain supported). @oscardssmith has access to the relevant packages. So I'll leave it to him to make that change. For the sorted tree issue, probably we need a test case here that has a generic version of that. @oscardssmith can you put together an example that imposes a tree structure onto a unit range `1:n`? I.e. if you iterate over Leaves, the corresponding node values are ranges `1:i, (i+1:j), (j+1:k), ..., (l+1:n)`, with the node values of the higher nodes in the tree being the union of the nodevalues of their children. I think that would well approximate the structure of what we're doing. As long as Leaves cursor corresponding to the leaf whose range contains a particular index `k` can be constructed in `O(depth)` time, I think we're ok algorithmically.

**Comment by ExpandingMan** [comment]
I *think* I now have all functionality except for `ShadowTree` and that bizarre `searchsortedlastmethod`.

I just spent a bit of time looking into `ShadowTree`.  It doesn't make a whole lot of sense to me, but I think it was mainly intended to do the sort of thing that I now allow for with `treemap` and `MapNode`.  I did a search with Juliahub if this is being used anywhere, and the only thing I found was [here](https://github.com/fabid/BehaviorTree.jl/blob/master/src/visualization.jl) which doesn't exactly do much to justify the existence of `ShadowTree` to me.

As of writing, I fully intend to drop `ShadowTree` altogether as I believe the functionality is replaced by `MapNode` (perhaps with `IndexNode`).  If someone has an example of something that can be done easily with `ShadowTree` that cannot be done with `MapNode` please provide.

Another change that I could imagine objections to is that my version no longer has a `treemap!` that provide in-place mutation, at least not directly (it is still possible to write a function that mutates the past nodes, it just cannot alter the overall structure of the existing tree).  The original case was quite specialized to the case where all non-leaf nodes in the tree are arrays.  I looked for cases of `treemap!` being used, again [in Juliahub](https://juliahub.com/ui/Search?q=treemap%21&type=code) but I found literally not a single case (the `treemap!` in HierarchicalUtils is an entirely different function).

**Comment by oscardssmith** [comment]
I think this is good to merge.

**Comment by ExpandingMan** [comment]
Cool.  I'd be quite happy if we merged this, but work remains before it can be tagged.  More complete unit testing is probably the most important item but I expect we'll want to make a few changes based on my ecosystem analysis (I'm running through packages now, looking pretty good so far).

Please merge if you want, and I'll continue my work from the new master.

**Comment by ExpandingMan** [comment]
Btw, I just ran unit tests for Flux (a direct dependency) to see if there was any breakage and no, there is not.  Don't know how significant that is because I frankly have no idea how Flux uses this, but it was gratifying at least.

**Comment by jlumpe** [comment]
I'm sorry, I wrote a bunch of comments but forgot to submit the review. I've done it now just for reference.

File: src/indexing.jl [hunk 24]
```diff
@@ -1,89 +1,132 @@
...
-Tree(x::AbstractShadowTree) = x
-show(io::IO, tree::Tree) = print_tree(io, tree.x)
+Obtain a node from a tree with [`IndexedChildren`](@ref) by indexing each level of the tree with the elements
+of `idx`.
 
-mutable struct AnnotationNode{T}
-    val::T
-    children::Vector{AnnotationNode{T}}
-end
+Note that this is a separate concept form indexed trees which by default do not have `IndexedChildren()`,
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/indexing.jl:8]: fixed

File: docs/src/faq.md [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+    children's *iteration states*.  To make matters worse, Julia's `Base` does not include traits
+    for describing these, and the `Base` iteration protocol makes very few assumptions about them.
+
+All of this means that you are unlikely to get type-stable code from AbstractTrees.jl without some
+effort.
+
+The simplest way around this is to define the `eltype` of tree iterators
+```julia
+Base.IteratorEltype(::Type{<:TreeIterator{ExampleNode}}) = Base.HasEltype()
+Base.eltype(::Type{<:TreeIterator{ExampleNode}}) = ExampleNode
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply docs/src/faq.md:85]: I'm not really sure what you mean here... yes, defining `eltype` for `TreeIterator` gives a guarantee for *all* nodes in a tree.  I made the conscious decision not to handle typed sub-trees because it just seems too complicated.  Those who want type stable code need trees with all nodes of the same type, I think we can live with this, the alternatives are horrendous.

File: docs/src/index.md [hunk 26]
```diff
@@ -1,31 +1,122 @@
...
+allow nodes to be accessed from a centralized tree object.
+
+```@docs
+IndexNode
+```
+
+### Functions
+All indexed trees *must* implement [`childindices`](@ref).
+
+Indexed trees rely on [`nodevalue(::Any, ::Any)`](@ref) for obtaining the value of a
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply docs/src/index.md:101]: Not sure if Documenter follows the ref in this case, and regardless I'd rather not change it in case the function arguments change (don't know if that would break Documenter).

File: src/base.jl [hunk 27]
```diff
@@ -1,59 +1,149 @@
-#
-# children() and related functions
-#
+
+"""
+    nodevalue(node)
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/base.jl:3]: That's a really good point... in the back of my mind I was thinking `nodevalue.(Iterator(x))` but of course that doesn't work because it assumes an array... it's pretty simple to just do something like `Iterators.map(nodevalue, Iterator(x))` but that's so verbose for something so simple.  I guess this is not the proudest moment for Julia syntax.

I'm going to think about this a bit more... off the top of my head `Iterator(nodevalue, x)` is the best I can come up with.

File: src/indexing.jl [hunk 28]
```diff
@@ -1,89 +1,132 @@
-# Wrappers which allow for tree-like indexing into objects
 
+"""
+    childindex(node, idx)
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/indexing.jl:3]: Excellent!  Will change.

File: src/traits.jl [hunk 29]
```diff
@@ -1,57 +1,171 @@
...
-struct RegularTree <: TreeKind; end
-struct IndexedTree <: TreeKind; end
+"""
+    ChildIndexing(::Type{N})
+    ChildIndexing(node)
 
-treekind(tree::Type) = RegularTree()
-treekind(tree) = treekind(typeof(tree))
+A trait indicating whether the tree node `n` has children (as returned by [`children`](@ref)) which can be
+indexed using 1-based indexing.
```
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/traits.jl:95]: No, it's not, I was just being lazy (or only trying to pack so many things into this one PR).

It might be worth considering how this can change before we tag 0.4.  Certainly the generalization to non 1-based indices is straightforward, but I don't think we care very much about that, and it could be added without breaking.  Less sure how to deal with dictionaries... however since both arrays and dictionaries define `keys` maybe there's hope.

File: src/base.jl (hunk 27 above)
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/base.jl:3]: I didn't do `Iterator(nodevalue, x)` because `PreOrderDFS` already has a function predicate, which is a bit unfortunate.  I instead added a method of `Base.values` which I hope isn't *too* egregious an abuse of that function...

File: src/traits.jl (hunk 29 above)
[the remark this answers is not in the document]
Comment (ExpandingMan) [reply src/traits.jl:95]: See discussion [here](https://github.com/JuliaCollections/AbstractTrees.jl/issues/98).

Some of what's going on here is that `getdescendant` (thanks for the name! way better than `childindex`) depends on an unnamed trait, not `ChildIndexing`.  I've decided not to create a trait for this, since I don't think it will be very useful, but internally `IndexedChildren` is only beneficial because it can assume ordered integers as the indices, so this will stay as is.

Happily, you *CAN* use `getdescendant` even for `NonIndexedChildren` and even for completely different indexing types like dictionaries, I've added some documentation to make this clear.

File: docs/src/faq.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (jlumpe) [reply docs/src/faq.md:85]: The trait would be to indicate that all nodes in the tree are of the same type. I just think there's a use for this property in several other cases besides just in `TreeIterator`. If defined this would have to be the return value of `parent()` (or `nothing`), `getdescendant()`, etc. and also the eltype of `children()`.

[the remark this answers is not in the document]
Comment (ExpandingMan) [reply docs/src/faq.md:85]: Good point!  I will change it later.
